typesetters/base.lua

```diff
@@ -912,6 +947,162 @@ function typesetter.leadingFor (_, vbox, previous)
+-- use one alone and break the balance of the Force.
+local linerEnterNode = pl.class(SILE.nodefactory.zerohbox)
+function linerEnterNode:_init(name, outputMethod)
+  SILE.nodefactory.hbox._init(self)
+  self.outputMethod = outputMethod
+  self.name = name
+  self.is_enter = true
+end
+function linerEnterNode:clone()
+  return linerEnterNode(self.name, self.outputMethod)
+end
+function linerEnterNode:__tostring ()
+  return "+L[" .. self.name .. "]"
+end
+local linerLeaveNode = pl.class(SILE.nodefactory.zerohbox)
+function linerLeaveNode:_init(name)
+  SILE.nodefactory.hbox._init(self)
+  self.name = name
+  self.is_leave = true
+end
+function linerLeaveNode:clone()
+  return linerLeaveNode(self.name)
+end
+function linerLeaveNode:__tostring ()
+  return "-L[" .. self.name .. "]"
+end
+
+local linerBox = pl.class(SILE.nodefactory.hbox)
+function linerBox:_init (name, outputMethod)
+  SILE.nodefactory.hbox._init(self)
+  self.width = SILE.length()
+  self.height = SILE.length()
+  self.depth = SILE.length()
+  self.name = name
+  self.inner = {}
+  self.outputYourself = outputMethod
+end
+function linerBox:append (node)
+  self.inner[#self.inner+1] = node
+  self.width = self.width + node.width
+  self.height = SU.max(self.height, node.height)
+  self.depth = SU.max(self.depth, node.depth)
+end
+function linerBox:outputContent (tsetter, line)
+  for _, node in ipairs(self.inner) do
+    node.outputYourself(node, tsetter, line)
+  end
+end
+function linerBox:__tostring ()
+  return "*L[" .. self.name .. "]H<" .. tostring(self.width) .. ">^" .. tostring(self.height) .. "-" .. tostring(self.depth) .. "v"
+end
+
+--- Any unclosed liner is reopened on the current line, so we clone and repeat
+-- it.
+-- An assumption is that the inserts are done after the current slice content,
+-- supposed to be just before meaningful (visible) content.
+---@param slice   table   Current line nodes
+---@return boolean        Whether a liner was reopened
+function typesetter:repeatEnterLiners (slice)
+  local m = self.state.liners
+  if #m > 0 then
+    for i = 1, #m  do
+      local n = m[i]:clone()
+      slice[#slice+1] = n
+      SU.debug("typesetter.liner", "Reopening liner", n)
+    end
+    return true
+  end
+  return false
+end
+
+--- All pairs of liners are rebuilt as hboxes wrapping their content.
+-- Migrating content, however, must be kept outside the hboxes at top slice level.
+---@param  slice  table   Flat nodes from current line
+---@return table          New reboxed slice
+function typesetter.reboxLiners (_, slice)
+  local outSlice = {}
+  local migratingList = {}
+  local lboxStack = {}
+  for i = 1, #slice do
+    local node = slice[i]
+    if node.is_enter then
+      SU.debug("typesetter.liner", "Start reboxing", node)
+      local n = linerBox(node.name, node.outputMethod)
+      lboxStack[#lboxStack+1] = n
+    elseif node.is_leave then
+      if #lboxStack == 0 then
+        SU.error("Multiliner box stacking mismatch" .. node)
```

Comment:
SILE manual fails here
```
 Multiliner box stacking mismatch-L[color_1] at documentation/c02-gettingstarted.sil (...) near 34:3
```

To investigate.

Reply:
Offending scenario:

```
\begin[class=book]{document}
\use[module=packages.autodoc]
\set[parameter=autodoc.highlighting, value=true]

\begin[type=autodoc:codeblock]{raw}
\begin{document}
Hello SILE!
\end{document}
\end{raw}

\end{document}
```

Running with `-d typesetter.liner`, one can open a color, and break to the next line immediately, it's legit. And `lastContentNodeIndex` (comment below) is indeed ill-named lol.

Reply:
Fixed the one. Name kept, finally, but it had to be set correctly then after repeating entry liner nodes.